Initial implementation of fullfiles

cmarcelo opened the pull request:
Simple starting point: uses compress/gzip as the only option (but already assumes there is a list of options).

cmarcelo committed:
swupd: add a helper to convert ftype back to string

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo committed:
swupd: add CreateFullfiles

Create fullfiles from a given manifest and chroot. Code is ready to
evaluate multiple compression implementations falling back to
uncompressed tar if no compression works. For now only Gzip is
available (using compress/gzip package).

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo committed:
swupd: add a test program for manually create fullfiles

This program is used to manually test the creation of the fullfiles
until it is integrated with the mixer tool itself.

To build, use

    $ make swupd/cmd/create-fullfiles.build

or without using the Makefile with

    $ GOPATH=$PWD go build swupd/cmd/create-fullfiles

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo commented:
There are more of these coming, but it is in a state worthy of a review already.

What is already implemented locally but still need to be put in a proper commit:
- Using multiple goroutines to do the work.
- xz and bzip2 compressors: they are trivial but want to make sure we have our vendoring strategy in place. Will add the dependencies and then the compressors.

What is still to be done:
- Automated testing. Should I cook some dummy chroot and etc and put on our testdata/? Or maybe take some partial from a real one (like just take the first 100 files)?
- Adding external compressor using "tar" command. Is it worth adding?

cmarcelo commented:
The helper program depends on the fix that is part of #6.

matthewrsj commented:
> Automated testing. Should I cook some dummy chroot and etc and put on our testdata/?

This is what I've been doing locally.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,229 @@
+	}
+
+	best := ""
+	var bestSize int64 = math.MaxInt64
```
Why not set bestSize to fi.Size?

cmarcelo replied:
It could work but at the moment if all our compressors give something larger than the original file, that would cause a failure that we don't handle yet.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,229 @@
+			os.RemoveAll(candidate)
+		}
+	}
+
```
This doesn't seem to allow for an empty fullfileCompressors list. For tiny files sometimes no compression is best.

cmarcelo replied:
Before sending this PR I've experimented with a tar-only "compressor", and the tar-header overhead took away any advantage of not compressing the small files. My understanding is that tar is always necessary to keep the metadata unless we make some other change elsewhere.

icarus-sparry replied:
I expect that any compressor will manage to compress the tar header,  I agree that we need (at the moment) to keep the tar wrapper around the file for the meta data.

Notation. I am going to use file.tar.gz for the gzipped version of the tarfile containing file, although I know it will actually be called file.tar eventually.

My point is that we should be comparing file.tar with  file.tar.gz, file.tar.xz, and file.tar.bz, rather than just comparing the latter 3 among themselves  . If you have an empty list of compressors then file.tar is the winner as there are no better compressed versions (as there are no compressed versions at all). At the moment if you have an empty list of compressors you get an error for everything, whilst I assert that if you don't have any compressors the best you can do is use the file.tar as it is.

Does this make sense?

cmarcelo replied:
After some thought, I agree. Specially because we are getting rid of an "error scenario", since there is always a valid file to use.

I wasn't even considering "tar only" because of the tests I did (mostly the files from os-core, os-core-update, kernel-native) and because the current swupd-server doesn't do that. Since swupd-client does "tar xf" I think we will be fine having the tar only files.

I rewrote that part of the code a little bit, to reuse the uncompressed tar file (taking advantage of it being a special thing, and not a "no compression" compressor). This will not serve for cases we are calling tar externally, but helps a little bit for the Go-only compressors.

PS: I'm being all "hey there can be externals" because we haven't validated the Go implementation tar yet, but I'm confident we will make things work without using the external tar tool or some C tar library implementation.

cmarcelo commented:
Changed the "swupd: add CreateFullfiles" patch after Icarus feedback, and added the patches for XZ and BZip2.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return fmt.Errorf("couldn't archive the file %s: %s", input, err)
+	}
+	uncompressedSize, err := uncompressed.Seek(0, io.SeekEnd)
```
My gut says we should do the size after the sync. OK, it just changes the error message.

cmarcelo replied:
I'll reorder, but my expectation is that Seek would work even without the Sync.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return fmt.Errorf("couldn't sync uncompressed fullfile: %s", err)
+	}
+	defer uncompressed.Close()
```
Why defer the close? Actually why Sync, why not just close?

Ah, I see, you are keeping the file open to pass to the compressors. Not sure this is a great idea, if we run the compressors in multiple go routines then the shared input pointer is going to be conflicted. Works OK at the moment but something to be aware of for the future.

cmarcelo replied:
Locally I've parallelized the execution at the fullfile level, not at the compressor choice level. IOW we would have a goroutines dealing with different input files. In that scenario, this would work OK.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,276 @@
+		return err
+	}
+	in.Close()
+	return tw.Close()
```
tarRegular ought to make an explicit commitment about the state of iostreams after it finishes. If you have errors then tw.Close is not called, so you leak a descriptor.

cmarcelo replied:
It is up to the caller to Close the 'w', which in this case is backed by a file descriptor. 'tw' is not backed by a file descriptor so not really important here.

The leak comment is relevant but for the caller of this function. I'll fix that.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,276 @@
+		return err
+	}
+	err = out.Close()
+	if err != nil {
```
Just return out.Close()

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,276 @@
+	return tar.FileInfoHeader(fi, "")
+}
+
+func compressGzip(args fullfileCompressArgs) (int64, error) {
```
We can probably abstract this a little more, after we have got things set up we want to

copy the data from an io.Reader to an io.Writer
get the "size" of the io.Writer
close the io.Writer
return the size
Do all of this with proper error handling.

For writing just the header, we can pass in a io.Discard (think /dev/null).

cmarcelo replied:
I did some factoring of redundant code by not trying to fit both external and internal compressors in the same bucket. I'll handle external compressors with a different function type.

I haven't tried hard to factor out the handling for non-regular files, but will revisit that later.

icarus-sparry commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,89 @@
+func main() {
+	log.SetFlags(0)
+
+	cpuProfile := flag.String("cpuprofile", "", "write CPU profile to a file")
```
Nice.

icarus-sparry commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,89 @@
+		defer pprof.StopCPUProfile()
+	}
+
+	flag.Parse()
```
Doesn't flag.Parse() need to go before line 33 to get *cpuProfile set?

cmarcelo replied:
Last minute copy pasta. Will fix this.

icarus-sparry commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	if *outputDir == "" {
+		tempDir, err := ioutil.TempDir("", "fullfiles-")
```
Do we want the first param to be "" (which will make it use something like /tmp, or "." to use the current directory.

cmarcelo replied:
In my testing, I wanted to avoid messing with the current directory, but don't have a strong opinion on the matter. I will change to use ".".

icarus-sparry commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,89 @@
+		*outputDir = tempDir
+	} else {
+		if _, err := os.Stat(*outputDir); err == nil {
+			log.Fatalf("output dir already exists, exiting to not override files")
```
s/override/overwrite/

icarus-sparry commented on src/swupd/README.vendor:
```diff
@@ -0,0 +1,25 @@
+This package uses govendor tool for "vendoring" packages,
```
Is "govendor" the right tool. The tool that is almost certainly going to be the eventual solution is [go dep](https://github.com/golang/dep)

cmarcelo replied:
govendor is sufficient. Not sure if it is "the right one". I'm not familiar with go dep, other than "when it is ready to use, it will be merged into go". I can take a look at it and follow up.

cmarcelo replied:
Following up: I read on go dep, and was convinced that we can start using it even if it is not merged to go tool yet. The workflow is similar to govendor for our use cases. I've updated the commits that add vendor code and the README.vendor.

icarus-sparry commented on src/swupd/fullfiles.go:
```diff
@@ -274,3 +277,31 @@ func compressGzip(args fullfileCompressArgs) (int64, error) {
+	if err != nil {
+		return 0, err
+	}
+	_, err = io.Copy(xw, args.tarredReader)
```
See previous comments about abstracting from here to the end of this function.

cmarcelo reviewed (commented):
I'll give another shot at factoring out some redundancy (after having the three cases).

cmarcelo commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,274 @@
+	}
+	// Sync the data so a compressor (possibly external program) can read this file while we
+	// still have an open file to pass to the compressors that read directly from it.
+	err = uncompressed.Sync()
```
We shouldn't really need this. I had in mind external tools reusing the tar we just wrote, but we don't have or plan to have that. I'll let a note around and if that ever happens we close/reopen or sync the file.

cmarcelo committed:
swupd: also use external tar+compressors in fullfile generation

The original swupd-server uses tar and the compressors called from it,
so keeping this around until we decide the Go's tar and compressors
implementation is good to use in production.

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo committed:
swupd: parallelize fullfile generation

Distribute each input file to the goroutine pool of task runners.

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

matthewrsj commented:
@cmarcelo is this in a reviewable state?

cmarcelo commented:
Yes. :-)

The only thing missing is some kind of automated test, but I don't expect it to change much the patches already published.

matthewrsj commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,88 @@
+		log.Fatalf("couldn't access the full manifest: %s", err)
+	}
+
+	// TODO: Make ReadManifestFromFile return the Manifest.
```
good idea.

matthewrsj commented on src/swupd/cmd/create-fullfiles/main.go:
```diff
@@ -0,0 +1,88 @@
+	os.Exit(2)
+}
+
+func main() {
```
Why make this its own program? I thought we were leaving this as a library for mixer to consume, or were you thinking this would be Exec'd by mixer?

You have CreateFullfiles below, but this function does a lot of input validation that will need to go in mixer itself if we are using CreateFullfiles as a library function. I guess I'm just wondering what the purpose of making this its own program is. Is it an example? For testing?

cmarcelo replied:
It is just a test program (maybe GitHub interface made it hard to see but the commit message tells that ;-). The functionality I've implemented is in the library function.

There isn't much real work done by it, most of the code is just to make sure the directories passed to it are correct. Some of that later will be done by Mixer (or one other library function) when it calls CreateFullfiles.

matthewrsj replied:
Ah, I reviewed the entire PR, not commit at a time so I missed it.

matthewrsj commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,345 @@
+		var candidateSize int64
+		candidate := fmt.Sprintf("%s.%d.%s", output, i, c.Name)
+
+		if c.Func != nil {
```
Ah, nice way to handle execs vs library calls.

matthewrsj reviewed (commented):
The code LGTM, I just had a question. Also, some of these functions could use unit tests.

cmarcelo committed:
swupd: add a simple test for CreateFullfiles

This test covers the basic end-to-end of CreateFullfiles. There are
several points of improvement, but it guarantees that at least it does
something.

Note that we can use "invalid" hashes because CreateFullfiles doesn't
validate them.

Signed-off-by: Caio Marcelo de Oliveira Filho <caio.oliveira@intel.com>

cmarcelo commented:
Added a simple test case.

matthewrsj commented:
This is good for now, some more exhaustive testing would be nice in the future, but this test is at least reasonable and passes. I'm going to merge it.

pohly commented on src/swupd/fullfiles.go:
```diff
@@ -0,0 +1,345 @@
+	var buf bytes.Buffer
+	// We know that name is suitable to be part of the transform regexp. If this function gets
+	// wider usage, we must revisit this.
+	cmd := exec.Command("tar", "--create", extra, "--file="+output, "--transform=s/.*/"+name+"/", "--preserve-permissions", input)
```
You must tell tar to use numeric IDs, otherwise unpacking of a tar archive with a user that isn't defined on the client (yet, because it gets created as part of the update) will fail. Same for using the internal tar module.

I'm a bit worried that mixer now ends up creating the actual tar archive in two different ways: with the internal tar module or with the external tar command, just compressed differently. That's likely to give you slightly different results, which could become a source of future bugs. Wouldn't it be better to create the tar content internally, perhaps even buffered in memory, and then just compress it in different ways?

Has someone verified that the Go tar module handles xattrs correctly and in a way that is compatible with swupd-client, which now uses libarchive?

The goal was to not rely on external commands (see discussion around using mixer with Yocto when it runs under pseudo, aka https://github.com/clearlinux/swupd-server/pull/48). Shall I create an issue in this repo to track that?

cmarcelo replied:
> You must tell tar to use numeric IDs, otherwise unpacking of a tar archive with a user that isn't defined on the client (yet, because it gets created as part of the update) will fail.

You mean passing --numeric-owner?

> Same for using the internal tar module.

The internal tar module should always include the numeric IDs.

> I'm a bit worried that mixer now ends up creating the actual tar archive in two different ways (...)

From the commit message:

    The original swupd-server uses tar and the compressors called from it,
    so keeping this around until we decide the Go's tar and compressors
    implementation is good to use in production.

I've also used to compare to see if the Go compressors was producing good enough file sizes. I should've added more clear comment about that. I plan to resolve this before this gets used by mixer.

> Has someone verified that the Go tar module handles xattrs correctly and in a way that is compatible with swupd-client, which now uses libarchive?

I haven't yet, but I've added this to my TODO list to check if the files we generate are good enough for swupd-client. Is there a list of specific xattrs you care about or you rely on propagating all of them?

> The goal was to not rely on external commands (...)

I'm moving toward that goal as well, I'll try to be more verbose in the commentary about that in the future.

pohly commented:
On Wed, 2017-12-20 at 12:56 -0800, Caio Marcelo de Oliveira Filho
wrote:
> You mean passing --numeric-owner?

Yes. That's what pack.c did in swupd-server, too.

> From the commit message:
> The original swupd-server uses tar and the compressors called from
> it,
> so keeping this around until we decide the Go's tar and compressors
> implementation is good to use in production.
> I've also used to compare to see if the Go compressors was producing
> good enough file sizes. I should've added more clear comment about
> that. I plan to resolve this before this gets used by mixer.

Thanks for clarifying. I was indeed only looking at the code and the
review comments, not the commit message - my bad.

> I haven't yet, but I've added this to my TODO list to check if the
> files we generate are good enough for swupd-client. Is there a list
> of specific xattrs you care about or you rely on propagating all of
> them?

All of them.

cmarcelo commented:
> Yes. That's what pack.c did in swupd-server, too.

I missed that because this work was to tar the fullfiles. I'm moving to packs and will pick this up.

By the way, it seems a bug to not have this flag in the fullfile generation on the original swupd-server.
